File: benchmarks/livermore/k03_dot.f
C     Kernel 3 -- Inner product (dot product)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K03
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION Z(1001), X(1001)
C
C     Initialize arrays using SIGNEL pattern
C
      SCALED = 0.1D0
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 10 K = 1, 1001
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          Z(K) = (BUZZ - FIZZ) * SCALED
   10 CONTINUE
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 20 K = 1, 1001
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          X(K) = (BUZZ - FIZZ) * SCALED
   20 CONTINUE
C
C     Kernel loop
C
      N = 1001
      DO 100 REP = 1, 10000
          Q = 0.0D0
          DO 90 K = 1, N
              Q = Q + Z(K) * X(K)
   90     CONTINUE
  100 CONTINUE
C
      WRITE(*,900) Q
  900 FORMAT('K3 inner product: q = ', E25.15)
      END

File: benchmarks/livermore/k03_dot.c
/* K3 — Inner product (1-based indexing, matches Fortran exactly)
   Fortran: DIMENSION Z(1001), X(1001), N=1001
   DO K=1,N: Q = Q + Z(K)*X(K) */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     1001
#define NREPS 31726000

int main(void) {
    double x[1002], z[1002];

    signel(z+1, 1001);
    signel(x+1, 1001);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    double q = 0.0;
    for (long rep = 0; rep < NREPS; rep++) {
        q = 0.0;
        for (long k = 1; k <= N; k++) {
            q = q + z[k] * x[k];
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("q = %f\n", q);
    return 0;
}
